### PR TITLE
[Sprint: 50] XD-2139 - Add ftp sink module

### PR DESCRIPTION
**Pull request by fmarchand**
I have signed and agree to the terms of the SpringSource Individual
Contributor License Agreement.

This is almost the same as XD-2942 but for a sink module.

Tell me If I forgot something.

### CODE REVIEW

File: spring-xd-dirt/src/main/java/org/springframework/xd/dirt/modules/metadata/FtpSinkOptionsMetadata.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+ * Module options for FTP sink module.
+ *
+ * @author Franck Marchand
+ */
+@Mixin({ FtpConnectionMixin.class })
+public class FtpSinkOptionsMetadata {
+
+    private int clientMode = 0;
+
+    private String charset = "UTF-8";
```
Comment (ericbottard) [inline spring-xd-dirt/src/main/java/org/springframework/xd/dirt/modules/metadata/FtpSinkOptionsMetadata.java]: I don't see where this is used (and it shouldn't, actually :) )

File: spring-xd-dirt/src/main/java/org/springframework/xd/dirt/modules/metadata/FtpSinkOptionsMetadata.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+    private String mode = "REPLACE";
+
+    private boolean useTemporaryFilename = true;
+
+    public boolean isUseTemporaryFilename() {
+        return useTemporaryFilename;
+    }
+
+    @ModuleOption("allows you to suppress using a temporary file name while writing the file")
```
Comment (ericbottard) [inline spring-xd-dirt/src/main/java/org/springframework/xd/dirt/modules/metadata/FtpSinkOptionsMetadata.java]: I'd remove the "you" reference in the sentence.
Also, the description is negated compared to the option name and meaning, this is a bit misleading. I'd rephrase the other way around

File: spring-xd-dirt/src/main/java/org/springframework/xd/dirt/modules/metadata/FtpSinkOptionsMetadata.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    public void setUseTemporaryFilename(boolean useTemporaryFilename) {
+        this.useTemporaryFilename = useTemporaryFilename;
+    }
+
+    @NotBlank
+    public String getMode() {
+        return mode;
+    }
+
+    @ModuleOption("what to do if the file already exists (enum FileExistsMode)")
```
Comment (ericbottard) [inline spring-xd-dirt/src/main/java/org/springframework/xd/dirt/modules/metadata/FtpSinkOptionsMetadata.java]: remove the reference to enum, this will be taken care of automatically. (see above)

File: spring-xd-dirt/src/main/java/org/springframework/xd/dirt/modules/metadata/FtpSinkOptionsMetadata.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+    private String temporaryRemoteDir = "/";
+
+    private boolean autoCreateDir = true;
+
+    private String remoteFileSeparator = "/";
+
+    private String tmpFileSuffix = ".tmp";
+
+    private String mode = "REPLACE";
```
Comment (ericbottard) [inline spring-xd-dirt/src/main/java/org/springframework/xd/dirt/modules/metadata/FtpSinkOptionsMetadata.java]: Can we introduce an enum here?

File: spring-xd-shell/src/test/java/org/springframework/xd/shell/command/AbstractStreamIntegrationTest.java [hunk 5]
```diff
@@ -122,6 +123,12 @@ protected FtpSource newFtpSource() {
 		return ftpSource;
 	}
 
+    protected FtpSink newFtpSink() {
```
Comment (ericbottard) [inline spring-xd-shell/src/test/java/org/springframework/xd/shell/command/AbstractStreamIntegrationTest.java]: indentation seems to be off.

File: spring-xd-test-fixtures/src/main/java/org/springframework/xd/test/fixtures/FtpSink.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+		catch (IOException e) {
+			throw new AssertionError("Deletion of directory should have worked", e);
+		}
+	}
+
+	public File getRemoteServerDirectory() {
+		return remoteServerDirectory;
+	}
+
+	private class DummyUserManager extends AbstractUserManager {
```
Comment (ericbottard) [inline spring-xd-test-fixtures/src/main/java/org/springframework/xd/test/fixtures/FtpSink.java]: Maybe extract as a toplevel class, now that it is shared with the source

File: spring-xd-test-fixtures/src/main/java/org/springframework/xd/test/fixtures/FtpSink.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+import org.apache.ftpserver.usermanager.impl.WritePermission;
+import org.springframework.xd.test.fixtures.util.AvailableSocketPorts;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.Arrays;
+
+/**
+ * A fixture that helps testing the ftp sink. Creates a local FTP server and exposes a directory in order to be able
```
Comment (ericbottard) [inline spring-xd-test-fixtures/src/main/java/org/springframework/xd/test/fixtures/FtpSink.java]: The second sentence is unclear IMO, as the exposed File directory is not used to **push** files but quite the opposite: it's used to witness the fact that files have appeared

File: src/docs/asciidoc/Sinks.asciidoc [hunk 8]
```diff
@@ -119,6 +121,54 @@ $$name$$:: $$filename pattern to use$$ *($$String$$, default: `<stream name>`)*
...
+We use the file source, so create a file :
+
+[source,bash]
+----
+$ echo hello > /tmp/xd/input/mystream/test.txt
+----
+
+You should see on the ftp server side the file 'test.txt" with the content 'hello'.
+
+To pass the filename to the module you can use the header 'file_name' with the filename you wish to be used.
```
Comment (ericbottard) [inline src/docs/asciidoc/Sinks.asciidoc]: use backticks around file_name here

File: src/docs/asciidoc/Sinks.asciidoc [hunk 9]
```diff
@@ -119,6 +121,54 @@ $$name$$:: $$filename pattern to use$$ *($$String$$, default: `<stream name>`)*
...
+
+NOTE : From spring integration dorcumentation
+By default Spring Integration will use o.s.i.file.DefaultFileNameGenerator if none is specified. DefaultFileNameGenerator will determine the file name based on the value of the file_name header (if it exists) in the MessageHeaders, or if the payload of the Message is already a java.io.File, then it will use the original name of that file.
```
Comment (ericbottard) [inline src/docs/asciidoc/Sinks.asciidoc]: and here

**Comment by ericbottard** [comment]
Thanks for this awesome contribution Franck!
Overall, very few remarks. Let me know if you can address them, otherwise I'll do

File: spring-xd-test-fixtures/src/main/java/org/springframework/xd/test/fixtures/FtpSink.java (hunk 6 above)
Comment (fmarchand) [inline spring-xd-test-fixtures/src/main/java/org/springframework/xd/test/fixtures/FtpSink.java]: Yes that would be wiser :)

**Comment by fmarchand** [comment]
Thank you !
I'll try to fix these things tonight :)

File: spring-xd-dirt/src/main/java/org/springframework/xd/dirt/modules/metadata/FtpSinkOptionsMetadata.java (hunk 4 above)
Comment (fmarchand) [inline spring-xd-dirt/src/main/java/org/springframework/xd/dirt/modules/metadata/FtpSinkOptionsMetadata.java]: I'll use Enum FileSinkOptionsMetadata.Mode

File: spring-xd-dirt/src/main/java/org/springframework/xd/dirt/modules/metadata/FtpSinkOptionsMetadata.java (hunk 2 above)
Comment (fmarchand) [inline spring-xd-dirt/src/main/java/org/springframework/xd/dirt/modules/metadata/FtpSinkOptionsMetadata.java]: I used the description in [spring-integration-file-4.1.xsd](http://www.springframework.org/schema/integration/file/spring-integration-file-4.1.xsd). I rephrased it.

**Comment by fmarchand** [comment]
Job done :) I didn't know where to put FtpDummyUserManager class, so I put it at the same level as FtpSink fixture. I let you put it where it should be. I'm sure you know where that kind of stuff should be better than me :)

**Comment by markpollack** [comment]
Eric has been out the past 2 weeks, he will review when he returns.

**Comment by ghillert** [comment]
Merged with commit 1abedb192ddda6ea9a5c6fc2f657d9a3828a5ac2

Addressed a few things during merge:

* Addressed wording in `ftp.properties`
* Whitespace issues
* Polish Documentation
* Avoid Java 8 methods when using `Files.readAllLines`
